lib/core/CProgramCounters.cc

```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+#include <core/CProgramCounters.h>
+
+#include <core/CLogger.h>
+#include <core/CRapidJsonLineWriter.h>
+#include <core/CStatePersistInserter.h>
+#include <core/CStateRestoreTraverser.h>
+#include <core/CStringUtils.h>
+
+#include <rapidjson/document.h>
+#include <rapidjson/ostreamwrapper.h>
+
+#include <ostream>
+#include <string>
+
+namespace ml {
+namespace core {
+
+namespace {
+
+using TGenericLineWriter = core::CRapidJsonLineWriter<rapidjson::OStreamWrapper>;
+
+const std::string NAME_TYPE("name");
+const std::string DESCRIPTION_TYPE("description");
+const std::string COUNTER_TYPE("value");
+
+//! Persistence tags
+const std::string KEY_TAG("a");
+const std::string VALUE_TAG("b");
+
+//! Helper function to add a string/int pair to JSON writer
+void addStringInt(TGenericLineWriter& writer,
+                  const std::string& name,
+                  const std::string& description,
+                  uint64_t counter) {
+    writer.StartObject();
+
+    writer.String(NAME_TYPE);
+    writer.String(name);
+
+    writer.String(DESCRIPTION_TYPE);
+    writer.String(description);
+
+    writer.String(COUNTER_TYPE);
+    writer.Uint64(counter);
+
+    writer.EndObject();
+}
+}
+
+CProgramCounters::CProgramCounters()
+    : m_Counters(), m_Cache(), m_ProgramCounterTypes() {
+}
+
+CProgramCounters& CProgramCounters::instance() {
+    return ms_Instance;
+}
+
+CProgramCounters::TCounter& CProgramCounters::counter(counter_t::ECounterTypes counterType) {
+    if (counterType >= counter_t::E_LastEnumCounter) {
```

Comment:
You can still have values that don't correspond to a single enum value - if a scoped enum only has constants `e1 = 2` and `e2 = 4` then 6 will be considered valid even though it doesn't correspond to `e1` or `e2` and is greater than the highest single enum value.  So I don't think scoped enums completely solve this problem.

My suggestion to leave gaps for future expansion also makes this validation harder, so that's an argument against leaving gaps.

Reply:
what I meant was that scoped enums force you to be explicit and disallow some of the implicit int behavior, but they are not perfect, I agree.